File: app/src/main/java/closer/vlllage/com/closer/handler/share/SearchGroupsHeaderAdapter.java
package closer.vlllage.com.closer.handler.share;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.EditText;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.group.SearchGroupsAdapter;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.ui.RecyclerViewHeader;

public class SearchGroupsHeaderAdapter extends SearchGroupsAdapter {

    private static final int HEADER_VIEW_TYPE = -1;
    private final OnQueryChangedListener onQueryChangedListener;

    private RecyclerViewHeader header = new RecyclerViewHeader();

    private String headerText;

    public SearchGroupsHeaderAdapter(PoolMember poolMember, OnGroupClickListener onGroupClickListener, OnCreateGroupClickListener onCreateGroupClickListener, OnQueryChangedListener onQueryChangedListener) {
        super(poolMember, onGroupClickListener, onCreateGroupClickListener);
        this.onQueryChangedListener = onQueryChangedListener;
        setNoAnimation(true);
    }

    @NonNull
    @Override
    public SearchGroupsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case HEADER_VIEW_TYPE:
                return new SearchGroupsViewHolder(LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.list_text_header_item, parent, false));
        }

        return super.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull SearchGroupsAdapter.SearchGroupsViewHolder holder, int position) {
        if (position == 0) {
            holder.name.setText(headerText);
            EditText searchGroups = holder.itemView.findViewById(R.id.searchGroups);

            searchGroups.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    onQueryChangedListener.onQueryChanged(s.toString());
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });
        } else {
            super.onBindViewHolder(holder, position - 1);
        }

        header.onBind(holder, position);
    }

    @Override
    public void onViewRecycled(@NonNull SearchGroupsViewHolder holder) {
        super.onViewRecycled(holder);
        header.onRecycled(holder);
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        header.attach(recyclerView, $(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.feedPeekHeight) * 2);
    }

    @Override
    public int getItemViewType(int position) {
        return position == 0 ? HEADER_VIEW_TYPE : super.getItemViewType(position - 1);
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + 1;
    }

    public SearchGroupsHeaderAdapter setHeaderText(String headerText) {
        this.headerText = headerText;
        return this;
    }

    public interface OnQueryChangedListener {
        void onQueryChanged(String query);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/share/SearchGroupsHeaderAdapter.kt
package closer.vlllage.com.closer.handler.share

import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.group.SearchGroupsAdapter
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.ui.RecyclerViewHeader
import com.queatz.on.On

class SearchGroupsHeaderAdapter constructor(on: On,
                                onGroupClickListener: ((Group, View) -> Unit)?,
                                onCreateGroupClickListener: ((String) -> Unit)?,
                                private val onQueryChangedListener: OnQueryChangedListener)
    : SearchGroupsAdapter(on, false, onGroupClickListener, onCreateGroupClickListener) {

    private val header = RecyclerViewHeader()
    private var headerText: String? = null

    init {
        setNoAnimation(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            HEADER_VIEW_TYPE -> return SearchGroupsHeaderViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.list_text_header_item, parent, false))
        }

        return super.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position == 0) {
            when (holder) {
                is SearchGroupsHeaderViewHolder -> {
                    holder.name.text = headerText
                    holder.searchGroups.addTextChangedListener(object : TextWatcher {
                        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

                        }

                        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                            onQueryChangedListener.onQueryChanged(s.toString())
                        }

                        override fun afterTextChanged(s: Editable) {

                        }
                    })
                }
            }
        } else {
            super.onBindViewHolder(holder, position - 1)
        }

        header.onBind(holder, position)
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        super.onViewRecycled(holder)
        header.onRecycled(holder)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        header.attach(recyclerView, on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.feedPeekHeight) * 2)
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) HEADER_VIEW_TYPE else super.getItemViewType(position - 1)
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + 1
    }

    fun setHeaderText(headerText: String): SearchGroupsHeaderAdapter {
        this.headerText = headerText
        return this
    }

    interface OnQueryChangedListener {
        fun onQueryChanged(query: String)
    }

    companion object {
        private const val HEADER_VIEW_TYPE = -1
    }

    class SearchGroupsHeaderViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val name: TextView = view.findViewById(R.id.name)
        val searchGroups: EditText = view.findViewById(R.id.searchGroups)
    }

}
